### PR TITLE
feat(hack): generate template file

### PR DESCRIPTION
**Pull request by georgettica**
now we can
- use 'make generate-template-file' to generate it
- all commands of 'make generate' will do all of the generations needed

### CODE REVIEW

File: hack/configuration-anomaly-detection-template.Template.yaml [hunk 1]
```diff
@@ -0,0 +1,136 @@
+apiVersion: v1
+kind: Template
+parameters: []
+metadata:
+  name: configuration-anomaly-detection-template
+objects:
+- apiVersion: v1
```
Comment (shibumi) [inline hack/configuration-anomaly-detection-template.Template.yaml:7]: You can leave that namespace in the template, but the pipeline will ignore it. Namespaces are managed by App-SRE

File: hack/update-template/main.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	// marshal back into bytes
+	saasTemplateFileAsBytes, marshalBackErr := yaml.Marshal(&saasTemplateFile)
+	if marshalBackErr != nil {
+		panic(fmt.Errorf("could not marshal the bigMapping back: %v", marshalBackErr))
+	}
+	// print to stdout just so we know something happened and the code isn't broken
+	fmt.Printf("---\n%s\n---\n", string(saasTemplateFileAsBytes))
+	outputFile := filepath.Join(workingDirectory, "../configuration-anomaly-detection-template.Template.yaml")
```
Comment (shibumi) [inline hack/update-template/main.go:95]: The target directory for the created file should be openshift/template.yaml (that's what all projects are doing). We can change the path in the app-sre deployment spec, but then it will be different to all other projects :thinking: 

I  don't have a clear opinion about this.

File: hack/update-template/main.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		saasTemplateFile.Objects = append(saasTemplateFile.Objects, fileAsMap)
+	}
+
+	// marshal back into bytes
+	saasTemplateFileAsBytes, marshalBackErr := yaml.Marshal(&saasTemplateFile)
+	if marshalBackErr != nil {
+		panic(fmt.Errorf("could not marshal the bigMapping back: %v", marshalBackErr))
+	}
+	// print to stdout just so we know something happened and the code isn't broken
+	fmt.Printf("---\n%s\n---\n", string(saasTemplateFileAsBytes))
```
Comment (shibumi) [inline hack/update-template/main.go:94]: I think this can be shortened to: `---\n%s\n`
YAML has just "start" parameter and no end parameter as far as I know

File: hack/update-template/main.go (hunk 2 above)
Comment (georgettica) [reply hack/update-template/main.go:95]: Give the path and I will do it :)

**Comment by georgettica** [comment]
I'll check the validate tomorrow

File: hack/update-template/main.go (hunk 2 above)
Comment (shibumi) [reply hack/update-template/main.go:95]: See my other PR for the path.. it's really just <project root>/openshift/template.yaml as specified here: https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/38300/diffs#51d44de8da2934b1e8168e02e247073fe65f428f_0_41

Line 41

File: hack/update-template/main.go (hunk 3 above)
Comment (shibumi) [reply hack/update-template/main.go:94]: I don't see the "---" in your generated template file at all :thinking:  are you sure that your code works? do you want to add some tests to it?

Comment (georgettica) [reply hack/update-template/main.go:94]: It's not required :/
I just posted the yaml files.
I will remove the --- from the print.

Testing might be odd here as I don't know what I am checking.

This was supposed to be a bit better than a big bash script/python code

Comment (shibumi) [reply hack/update-template/main.go:94]: True, Go is the better solution for yes :)

I just didn't know that the "---"s  are not necessary :thinking: I thought we need them to differ between files/resources.
Did you try executing the YAML on a cluster? did it work?

Comment (georgettica) [reply hack/update-template/main.go:94]: No, but feel free to check

**Comment by maorfr** [comment]
the openshift template should assume it is deploying only what constitutes the "service", and not a full from-zero-to-hero bundle.

hence, you can remove the Namespace, the Subscription, etc.

**Comment by maorfr** [comment]
created [APPSRE-5497](https://issues.redhat.com/browse/APPSRE-5497) to take a deep dive

**Comment by georgettica** [comment]
might be related to APPSRE-5497

File: openshift/template.yaml [hunk 4]
```diff
@@ -0,0 +1,251 @@
+apiVersion: v1
+kind: Template
```
Comment (shibumi) [inline openshift/template.yaml:2]: Is this an Openshift API resource? :thinking:  never heard of template on Kubernetes. If yes, we might have to modify my saas file PR on gitlab, because I didn't specify that API resource there yet.

Comment (georgettica) [reply openshift/template.yaml:2]: nope, no need. this is the file they process.

**Comment by shibumi** [comment]
feel free to merge

**Comment by georgettica** [comment]
I need to test the template

File: openshift/template.yaml [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+      - name: CSS_JUMPROLE
+        value: XXXX
+      - name: SUPPORT_JUMPROLE
+        value: XXXX
+      envFrom:
+      - secretRef:
+          name: cad-aws-credentials
+      - secretRef:
+          name: cad-ocm-client-secret
+      image: quay.io/app-sre/cadctl@sha256:22d2a957d935e883f45addd35acc87450ce71ea2f94c33f2df70ff36b015486a
```
Comment (maorfr) [inline openshift/template.yaml:250]: this should be parameterized

File: openshift/template.yaml [hunk 6]
```diff
@@ -0,0 +1,251 @@
+apiVersion: v1
+kind: Template
+metadata:
+  name: configuration-anomaly-detection-template
+parameters: []
+objects:
+- apiVersion: tekton.dev/v1beta1
+  kind: PipelineRun
```
Comment (maorfr) [inline openshift/template.yaml:8]: what does this do?

File: openshift/template.yaml (hunk 5 above)
Comment (georgettica) [reply openshift/template.yaml:250]: yes, but to which value?

File: openshift/template.yaml (hunk 6 above)
Comment (georgettica) [reply openshift/template.yaml:8]: if you mean the PipelineRun. it's for testing purposes, we should move it out as well (as it causes and error when you apply it)

**Comment by georgettica** [comment]
errors I currently see when I run the template by a normal backplane user
```
Error from server (Forbidden): error when creating "STDIN": roles.rbac.authorization.k8s.io "cad-eventlistener-role" is forbidden: user "$CURRENT_USER" (groups=["system:serviceaccounts" "${CURRENT_GROUP}" "system:authenticated"]) is attempting to grant RBAC permissions not currently held:
{APIGroups:["*"], Resources:["configmaps"], Verbs:["list" "watch" "get"]}
{APIGroups:["*"], Resources:["secrets"], Verbs:["list" "watch" "get"]}
{APIGroups:["*"], Resources:["serviceaccounts"], Verbs:["list" "watch" "get"]}
{APIGroups:["tekton.dev"], Resources:["*"], Verbs:["*"]}
{APIGroups:["triggers.tekton.dev"], Resources:["*"], Verbs:["*"]}
Error from server (Forbidden): error when creating "STDIN": clusterroles.rbac.authorization.k8s.io is forbidden: User "$CURRENT_USER" cannot create resource "clusterroles" in API group "rbac.authorization.k8s.io" at the cluster scope
Error from server (NotFound): error when creating "STDIN": roles.rbac.authorization.k8s.io "cad-eventlistener-role" not found
Error from server (Forbidden): error when creating "STDIN": clusterrolebindings.rbac.authorization.k8s.io is forbidden: User "$CURRENT_USER" cannot create resource "clusterrolebindings" in API group "rbac.authorization.k8s.io" at the cluster scope
```

**Comment by georgettica** [comment]
@maorfr I need to know the namespace.
when I run this locally using the admin perms, all passes aside of
```
Error from server (Invalid): error when creating "STDIN": ClusterRoleBinding.rbac.authorization.k8s.io "cad-eventlistener-clusterrolebinding" is invalid: subjects[0].namespace: Required value
```

File: openshift/template.yaml (hunk 5 above)
Comment (maorfr) [reply openshift/template.yaml:250]: https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/docs/app-sre/continuous-delivery-in-app-interface.md#automatically-generated-parameters

**Comment by maorfr** [comment]
add a parameter of `NAMESPACE_NAME` with a default value of `configuration-anomaly-detection` and override it from the saas file to the right name?

File: openshift/template.yaml (hunk 6 above)
Comment (shibumi) [reply openshift/template.yaml:8]: This is a test PipelineRun and should not be included.

File: openshift/template.yaml (hunk 5 above)
Comment (shibumi) [reply openshift/template.yaml:250]: Ah yes, I forgot about this one. We want to parametrize this and very likely we have to change the names for the secrets as well

**Comment by georgettica** [comment]
please review, or at least tell me in broad strokes what you think of this

File: deploy/serviceaccount.yaml [hunk 7]
```diff
@@ -69,7 +69,6 @@ roleRef:
   apiGroup: rbac.authorization.k8s.io
   kind: Role
   name: cad-eventlistener-role
-  namespace: configuration-anomaly-detection
```
Comment (fahlmant) [inline deploy/serviceaccount.yaml:72]: Was this removal on purpose?

File: hack/update-template/main.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+)
+
+func main() {
+	ex, getExecutableErr := os.Executable()
+	if getExecutableErr != nil {
+		panic(fmt.Errorf("could not get the Executable: %w", getExecutableErr))
+	}
+
+	workingDirectory := filepath.Dir(ex)
+	fmt.Printf("workingDirectory :: %s\n", workingDirectory)
```
Comment (fahlmant) [inline hack/update-template/main.go:45]: Any reason not to do `os.Getcwd()` here?

File: hack/update-template/main.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+	var filesNames []string
+	relativeDeployFolder := filepath.Join(workingDirectory, deployFolderPath)
```
[the remark this answers is not in the document]
Comment (georgettica) [reply hack/update-template/main.go:48]: It's starting the search from the binaries folder, then traverses up

File: hack/update-template/main.go (hunk 8 above)
Comment (georgettica) [reply hack/update-template/main.go:45]: Yeah, it is where the caller is, not where the script is

File: hack/update-template/main.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (georgettica) [reply hack/update-template/main.go:48]: These can be later parameterized if need be.
Let me know if that is necessary

**Comment by georgettica** [comment]
I'll fix this tomorrow, I expected the code to work

File: hack/update-template/main.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (georgettica) [reply hack/update-template/main.go:48]: No, it directs to the location of the binary

File: hack/update-template/main.go [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+			removeNestedField(fileAsMap, "metadata", "namespace")
+
+			kind, kindExists := fileAsMap["kind"]
+			if !kindExists {
+				panic(fmt.Errorf("kind doesn't exist"))
+			}
+
+			apiVersion, apiVersionExists := fileAsMap["apiVersion"]
+			if !apiVersionExists {
```
Comment (fahlmant) [inline hack/update-template/main.go:107]: This might be better rolled into a switch statement

File: openshift/template.yaml (hunk 5 above)
Comment (georgettica) [reply openshift/template.yaml]: I will keep this up as @ninabauer might change the secrets in her PR

File: openshift/template.yaml [hunk 11]
```diff
@@ -0,0 +1,242 @@
+apiVersion: template.openshift.io/v1
+kind: Template
+metadata:
+  name: configuration-anomaly-detection-template
+parameters:
+- name: IMAGE_DIGEST
+  value: sha256:22d2a957d935e883f45addd35acc87450ce71ea2f94c33f2df70ff36b015486a
```
Comment (shibumi) [inline openshift/template.yaml:7]: Question: These are only set when the parameter hasn't been set, correct?

File: openshift/template.yaml [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+        value: XXXX
+      - name: CAD_SILENT_POLICY
+        value: XXXX
+      - name: CSS_JUMPROLE
+        value: XXXX
+      - name: SUPPORT_JUMPROLE
+        value: XXXX
+      envFrom:
+      - secretRef:
+          name: cad-aws-credentials
```
Comment (shibumi) [inline openshift/template.yaml:238]: I am not 100% sure, but these secret names might be different on the clusters. @maorfr  is there any way to verify the secrets on the app-sre clusters?

File: hack/update-template/unstructured-modification.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+func nestedFieldNoCopy(obj map[interface{}]interface{}, fields ...string) (interface{}, bool, error) {
+	var val interface{} = obj
+
+	for _, field := range fields {
+		if val == nil {
+			return nil, false, nil
+		}
+		if m, ok := val.(map[interface{}]interface{}); ok {
+			val, ok = m[field]
```
Comment (shibumi) [inline hack/update-template/unstructured-modification.go:30]: What's the purpose of vail if you don't use it inside of the loop? :thinking:

File: hack/update-template/unstructured-modification.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+		}
+	}
+	return val, true, nil
+}
+
+func removeNestedField(obj map[interface{}]interface{}, fields ...string) {
+	m := obj
+	for _, field := range fields[:len(fields)-1] {
+		if x, ok := m[field].(map[interface{}]interface{}); ok {
+			m = x
```
Comment (shibumi) [inline hack/update-template/unstructured-modification.go:45]: Same here.. I don't understand how this loop works. Are you sure that the last line `delete...) shouldn't be in the loop?
Or would it make sense to break the loop if you've found the m?

File: hack/update-template/unstructured-modification.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
+package main
+
+// all of these have been pulled from k8s.io/apimachinery/pkg/apis/meta/v1/unstructured
+// this issue https://github.com/kubernetes/apimachinery/issues/138 explains why I needed to duplicate these funcs
```
Comment (shibumi) [inline hack/update-template/unstructured-modification.go:4]: The issue doesn't really explain why you can't just import the package. The issue explains a problem `with` the package, but it doesn't explain what you've changed or what the package does different to the code in this file.

If this is the same code like in apimachinery... why do we copy it?

File: hack/update-template/unstructured-modification.go (hunk 13 above)
Comment (shibumi) [reply hack/update-template/unstructured-modification.go:30]: Ah okay never mind. It's about the **last valid field**! :D

File: hack/update-template/unstructured-modification.go (hunk 14 above)
Comment (shibumi) [reply hack/update-template/unstructured-modification.go:45]: Ah okay never mind. It's about the **last valid field**! :D

File: openshift/template.yaml (hunk 11 above)
Comment (georgettica) [reply openshift/template.yaml:7]: yes

File: hack/update-template/unstructured-modification.go (hunk 15 above)
Comment (georgettica) [reply hack/update-template/unstructured-modification.go:4]: yeah, I will write it there aswell. good point

Comment (shibumi) [reply hack/update-template/unstructured-modification.go:4]: That is much better

File: openshift/template.yaml (hunk 12 above)
Comment (maorfr) [reply openshift/template.yaml:238]: see if this helps: https://gitlab.cee.redhat.com/service/app-interface#manage-secrets-via-app-interface-openshiftnamespace-1yml-using-vault

File: openshift/template.yaml (hunk 11 above)
Comment (maorfr) [inline openshift/template.yaml:7]: consider removing `value` and adding `required: true` instead?

similar to https://github.com/openshift/aws-account-operator/blob/0dc92d1e582f8f381f9a212b2edc4725394e2cb6/hack/olm-registry/olm-artifacts-template.yaml#L11-L12

File: openshift/template.yaml (hunk 12 above)
Comment (maorfr) [reply openshift/template.yaml:238]: just for my curiosity, what are these aws credentials?

File: openshift/template.yaml (hunk 11 above)
Comment (georgettica) [reply openshift/template.yaml:7]: nah, I want the usual dev to have the correct image (and hopefully the correct one).
currently this image cannot pull the correct credentials, so it's ok I think

Comment (maorfr) [reply openshift/template.yaml:7]: so are you planning to maintain the image in this template forever?

Comment (georgettica) [reply openshift/template.yaml:7]: I was hoping I could, is that ask too much?
I can remove if it is

Comment (maorfr) [reply openshift/template.yaml:7]: i don't think it's feasible. you also can't know the digest before the image is pushed. and that only happens after a merge. :chicken: :egg:

Comment (maorfr) [reply openshift/template.yaml:7]: i think that would be better, and also consistent with other operators.

File: openshift/template.yaml (hunk 12 above)
Comment (georgettica) [reply openshift/template.yaml:238]: These are generated using a an internal command to pull AWS creds.
For prod we have the AWS key and secret fields

File: openshift/template.yaml (hunk 11 above)
Comment (georgettica) [reply openshift/template.yaml:7]: #60
